// game/player.js
var Weapon = require('game/weapon');

var Player = function Player (id, body) {
    this.id = id;
    this.body = body;
    this.health = 100;
    this.max_health = 100;
    this.weapon = new Weapon(this);
    this.name = '';
    this.team = '';
    this.flag = null;
    this.relativeCursorX = 0;
    this.relativeCursorY = 0;
};

Player.prototype = {
    tick: function(elapsedTime, game) {
        this.body.tick(elapsedTime);
        this.weapon.tick(elapsedTime, game);
        var center = this.body.center();
        this.weapon.onMoveSight(game, center, {x:this.relativeCursorX+center.x, y:this.relativeCursorY+center.y});
        if(this.flag) {
            this.flag.body.x = this.body.x;
            this.flag.body.y = this.body.y - this.flag.body.height;
        }
    },

    takeFlag: function(flag) {
        this.flag = flag;
        this.flag.isTook = true;
        this.flag.body.isCollidable = false;
        this.flag.isAtSpawn = false;
    },

    dropFlag: function() {
        if(this.flag) {
            this.flag.body.isCollidable = true;
            this.flag.isTook = false;
            this.flag = null;
        }
    }
};

module.exports = Player;
