// game/weapon.js
var Body = require('physics/body');
var Projectile = require('game/projectile');

var Weapon = function Weapon(player) {
    this.reloadDelay = 100;
    this.m_currentReload = 0; //Need to be > reloadDelay to fire
    this.player_id = player.id;
    this.isFiring = false;
    this.src = null;
    this.dest = null;
};

Weapon.prototype = {
    tick: function(elapsedTime, game) {
        if(this.m_currentReload <= this.reloadDelay) {
            this.m_currentReload += elapsedTime;
        }

        if(this.m_currentReload >= this.reloadDelay && this.isFiring) {
            this.fire(game);
        }
    },

    onTriggerPulled: function() {
        this.isFiring = true;
    },

    onTriggerReleased: function() {
        this.isFiring = false;
    },

    onMoveSight: function(game, src, dest) {
        this.src = src;
        this.dest = dest;
    },

    fire: function(game) {
        this.m_currentReload = 0;
        var proj = new Projectile(this.src, this.dest, game.getPlayer(this.player_id));
        game.addProjectile(proj);
    }
};

module.exports = Weapon;
